What's New 109 (#2377)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/whats-new/2023/01/devtools-109.md

*  Snapshots of the DOM.

When saving performance profiles, heap snapshots, allocation timelines, or allocation sampling, you can now export a new `.devtools` file format.  When you then import the `.devtools` file, a new instance of DevTools opens, with the state of your **Elements**, **Console**, and **Sources** tools preserved.  These enhanced traces offer a powerful new way to collaborate and share the information in your DevTools.

To try the enhanced traces feature:

1. In DevTools, click the **Settings** (![Settings icon](../../../media/settings-gear-icon-light-theme.png)) button.

1. In the **Persistence** section of the **Preferences** settings page, select the **Export enhanced performance and memory traces** checkbox:

   ![Settings > Preferences > Persistence checkboxes](./devtools-109-images/enhanced-trace-settings.png)

1. If you want to preserve Console messages, JavaScript, or DOM snapshots, select the corresponding checkboxes.

1. Click the **Close** (**x**) button in DevTools **Settings**.

1. In the **Performance** tool, take a recording.

1. Click the **Save profile** (![Save profile icon](./devtools-109-images/save-profile-icon.png)) button:

>    ![Exporting a trace](./devtools-109-images/enhanced-trace-export.png)

Reviewer: ```suggestion
```
We can cut this image. Exporting with the **Export enhanced performance and memory traces** setting automatically defaults to the `.devtools` format so we don't need to include it

Author: removed png and line

> 1. In the **Save As** dialog, save the new `.devtools` file.  The profile is saved with a new `.devtools` file format.

Reviewer: We mention the new file format in the announcement, We don't need to say it again here.
```suggestion
1. In the **Save As** dialog, save the new `.devtools` file.
```

Author: removed sent.

**Importing:**

1. Import the `.devtools` file from within the **Performance** tool by clicking the **Load profile** (![Load profile icon](./devtools-109-images/load-profile-icon.png)) button:

   ![Importing a trace from within the Performance tool](./devtools-109-images/enhanced-trace-import.png)

1. A new DevTools window opens, containing a subset of the tools, including the **Performance** tool loaded with the profile that you just recorded.  The **Elements**, **Console**, and **Sources** tools are also pre-populated with their preserved state:

   ![Trace import console](./devtools-109-images/enhanced-trace-import-console.png)

> You can also export from the **Memory** tool with the new `.devtools` file format.  This is a new feature that the DevTools team is actively experimenting with, so please send us your feedback in our GitHub repo!  Post comments in [[Feedback] Enhanced Traces experiment #122](https://github.com/MicrosoftEdge/DevTools/issues/122).

Reviewer: ```suggestion
You can also export from the **Memory** tool with the new `.devtools` file format.  This is a new feature that the DevTools team is actively experimenting with, so please send us your feedback in [[Feedback] Enhanced Traces experiment #122](https://github.com/MicrosoftEdge/DevTools/issues/122).
```

Author: fixed locally

*  Messages logged to the Console.
*  The JavaScript that was running on the page at the time of recording.
*  Snapshots of the DOM.

When saving performance profiles, heap snapshots, allocation timelines, or allocation sampling, you can now export a new `.devtools` file format.  When you then import the `.devtools` file, a new instance of DevTools opens, with the state of your **Elements**, **Console**, and **Sources** tools preserved.  These enhanced traces offer a powerful new way to collaborate and share the information in your DevTools.

To try the enhanced traces feature:

1. In DevTools, click the **Settings** (![Settings icon](../../../media/settings-gear-icon-light-theme.png)) button.

1. In the **Persistence** section of the **Preferences** settings page, select the **Export enhanced performance and memory traces** checkbox:

   ![Settings > Preferences > Persistence checkboxes](./devtools-109-images/enhanced-trace-settings.png)

1. If you want to preserve Console messages, JavaScript, or DOM snapshots, select the corresponding checkboxes.

1. Click the **Close** (**x**) button in DevTools **Settings**.

> 1. In the **Performance** tool, take a recording.  Or, in the **Memory** tool, take a heap snapshot.

Reviewer: ```suggestion
1. In the **Performance** tool, take a recording.
```
Let's simplify these directions just to focus on the Performance tool for the announcement. Let's add a link at the end of the announcement to the docs in this PR: https://github.com/MicrosoftDocs/edge-developer/pull/2388 when it lands

Author: Removed text.  Draft PR link exists and "todo" is noted

> 1. If you use the **Performance** tool, right-click the **Save profile** (![Save profile icon](./devtools-109-images/save-profile-icon.png)) button, and then select **.devtools (enhanced format for Microsoft Edge)**:

Author: done

   ![Exporting a trace](./devtools-109-images/enhanced-trace-export.png)

1. In the **Save As** dialog, save the new `.devtools` file.

>    Or, in the **Memory** tool, click the **Save** link next to the heap snapshot.  In the **Save As** dialog, save the new `.devtools` file.

Reviewer: ```suggestion
```

Author: removed Memory tool scenario (this line)

**Importing:**

> 1. Import the `.devtools` file from within the **Performance** tool or **Memory** tool.  If you use the **Performance** tool, click the **Load profile** (![Load profile icon](./devtools-109-images/load-profile-icon.png)) button:

Reviewer: ```suggestion
1. Import the `.devtools` file from within the **Performance** tool. Click the **Load profile** (![Load profile icon](./devtools-109-images/load-profile-icon.png)) button:
```

Author: removed Memory tool scenario

   ![Importing a trace from within the Performance tool](./devtools-109-images/enhanced-trace-import.png)

>    If you use the **Memory** tool, right-click the **Heap Snapshots** pane and then select **Load**.

Author: removed

> 1. A new DevTools window opens, containing a subset of the tools, including the **Memory** or **Performance** tool.  The **Elements**, **Console**, and **Sources** tools are pre-populated with the preserved state:

Author: is "Memory or Performance tool" correct?

> 1. A new DevTools window opens, containing a subset of the tools, including the **Performance** or **Memory** tool.  The **Elements**, **Console**, and **Sources** tools are pre-populated with the preserved state:

Author: Is "or" correct?  
`including the **Performance** or **Memory** tool"`

Reviewer 2: No I think it's "and". Both tools appear in the new window.

Reviewer: It's both but in this example, only the Performance tool is populated.

```suggestion
1. A new DevTools window opens, containing a subset of the tools, including the **Performance** tool.  The **Elements**, **Console**, and **Sources** tools are pre-populated with the preserved state:
```

Author: trying and/or, eg:

A new DevTools window opens, containing a subset of the tools, including the **Performance** and/or **Memory** tool.

This is a new feature that the DevTools team is actively experimenting with, so please [send us your feedback in our GitHub repo](https://github.com/MicrosoftEdge/DevTools/issues)!

See also:
* [Share performance and memory traces (PR 2388 draft)](https://github.com/MicrosoftDocs/edge-developer/blob/user/pabrosse/enhanced-traces/microsoft-edge/devtools-guide-chromium/evaluate-performance/share-traces.md)
<!-- todo: link to final article when PR is live, after merge main into present branch.  expect:
* [Share performance and memory traces](../../../evaluate-performance/share-traces.md)
-->

<!-- ====================================================================== -->
## Debug long-running Recalculate Styles events in the Performance tool with new selector stats

<!-- Subtitle: Understand which of your CSS selectors are contributing to slow performance of your website or app. -->

In Microsoft Edge 109, in the **Performance** tool, you can turn on the **Enable advanced rendering instrumentation (slow)** setting in the **Performance** tool to access the new selector stats feature.

With this setting turned on:

> 1. Select **Record** and execute the scenario you want to improve on your website or app.

Author: @zoherghadyali - confirmed, https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/execute

1. Select **Stop**.

1. Select a **Recalculate Style** event.  In the bottom section of the **Performance** tool, select the **Selector Stats** tab:

   ![Performance tool Settings 'Enable advanced rendering instrumentation (slow)' checkbox, and 'Selector Stats' tab](./devtools-109-images/advanced-rendering-instrum.png)

>    The **Selector Stats** tab provides a list of all the CSS selectors that were calculated by the browser engine during the **Recalculate Style** event.  You can sort selectors by the **Elapsed** time they took to process, or by the number of elements they matched (the **Match Count** column).  Use this data to:

Reviewer: ```suggestion
The **Selector Stats** tab provides a list of all the CSS selectors that were calculated by the browser engine during the **Recalculate Style** event.  You can sort selectors by the **Elapsed** time they took to process, or by the number of elements they matched (the **Match Count** column).  Use this data to:
```
Feel like this was weirdly indented in the preview. Here's what I'm seeing: 
![image](https://user-images.githubusercontent.com/34316555/212152834-cd235b43-ad5d-448a-87f7-487fc0599987.png)

Author: changed

   * Find selectors that take a long time for the browser to process, and simplify them.
   * Make selectors more specific, to improve performance.

See also:
* The truth about CSS selector performance - pending blog post<!-- todo: incoming PR -->
* [[Feedback] Selector Performance Tracing Explainer · Issue #98](https://github.com/MicrosoftEdge/DevTools/issues/98)

> 

Reviewer: ```suggestion
* [Analyze selector performance during Recalculate Style events](https://learn.microsoft.com/microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats)
```

for when https://github.com/MicrosoftDocs/edge-developer/pull/2399 lands

Author: updated link.  In rendered, shows:
[Analyze selector performance during Recalculate Style events (PR 2399 draft)]
.md contains "todo: finalize link"

<!-- ====================================================================== -->

> ## Track objects discarded by garbage collection when Allocation sampling in the Memory tool

Reviewer: ```suggestion
## Track objects discarded by garbage collection when allocation sampling
```
Feel like this title is too long. I don't really care whether we capitalize the first A in "Allocation sampling" or not between this announcement and the docs but let's just be consistent about it. I'll defer to you

Author: fixed locally, lowercase

*  Messages logged to the Console.
*  The JavaScript that was running on the page at the time of recording.
*  Snapshots of the DOM.

When saving performance profiles, heap snapshots, allocation timelines, or allocation sampling, you can now export a new `.devtools` file format.  When you then import the `.devtools` file, a new instance of DevTools opens, with the state of your **Elements**, **Console**, and **Sources** tools preserved.  These enhanced traces offer a powerful new way to collaborate and share the information in your DevTools.

To try the enhanced traces feature:

1. In DevTools, click the **Settings** (![Settings icon](../../../media/settings-gear-icon-light-theme.png)) button.

1. In the **Persistence** section of the **Preferences** settings page, select the **Export enhanced performance and memory traces** checkbox:

   ![Settings > Preferences > Persistence checkboxes](./devtools-109-images/enhanced-trace-settings.png)

1. If you want to preserve Console messages, JavaScript, or DOM snapshots, select the corresponding checkboxes.

1. Click the **Close** (**x**) button in DevTools **Settings**.

1. In the **Performance** tool, take a recording.

> 1. Right-click the **Save profile** (![Save profile icon](./devtools-109-images/save-profile-icon.png)) button, and then select **.devtools (enhanced format for Microsoft Edge)**:

Reviewer: Per feedback from Rob, we can simplify this step. I opted to keep all the previous steps as well because I think they show the value prop of the feature.

```suggestion
1. Click the **Save profile** (![Save profile icon](./devtools-109-images/save-profile-icon.png)) button and note that the profile is being saved with a new `.devtools` file format.
```

Author: fixed locally

   ![Exporting a trace](./devtools-109-images/enhanced-trace-export.png)

1. In the **Save As** dialog, save the new `.devtools` file.

**Importing:**

1. Import the `.devtools` file from within the **Performance** tool by clicking the **Load profile** (![Load profile icon](./devtools-109-images/load-profile-icon.png)) button:

   ![Importing a trace from within the Performance tool](./devtools-109-images/enhanced-trace-import.png)

> 1. A new DevTools window opens, containing a subset of the tools, including the **Performance** and/or **Memory** tool.  The **Elements**, **Console**, and **Sources** tools are pre-populated with the preserved state:

Reviewer: ```suggestion
1. A new DevTools window opens, containing a subset of the tools, including the **Performance** tool loaded with the profile you just recorded.  The **Elements**, **Console**, and **Sources** tools are also pre-populated with their preserved state:
```

Author: done locally

   ![Trace import console](./devtools-109-images/enhanced-trace-import-console.png)

> This is a new feature that the DevTools team is actively experimenting with, so please [send us your feedback in our GitHub repo](https://github.com/MicrosoftEdge/DevTools/issues)!

Reviewer: ```suggestion
You can also export from the **Memory** tool with the new `.devtools` file format. This is a new feature that the DevTools team is actively experimenting with, so please [send us your feedback in our GitHub repo](https://github.com/MicrosoftEdge/DevTools/issues/122)!
```

Author: fixed locally, with link text matching destination title per style guides

<!-- todo: link to article when PR is live, after merge main into present branch
See also:

> * [Share performance and memory traces](../../../evaluate-performance/share-traces.md)

Reviewer: ```suggestion
* [Share performance and memory traces](../../../experimental-features/share-traces.md)
```

Author: not available yet

Reviewer: It is available now: https://learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/experimental-features/share-traces

Author: link is present
https://github.com/mikehoffms/edge-developer/blob/user/mikehoffms/whats-new-109/microsoft-edge/devtools-guide-chromium/whats-new/2023/01/devtools-109.md#import-and-export-your-devtools-instances-with-the-new-enhanced-traces-feature

PR that adds that article:
https://github.com/MicrosoftDocs/edge-developer/pull/2388 "Documentation for the DevTools enhanced traces feature"
Rendered:
https://github.com/MicrosoftDocs/edge-developer/blob/user/pabrosse/enhanced-traces/microsoft-edge/devtools-guide-chromium/evaluate-performance/share-traces.md
-->

<!-- ====================================================================== -->

> ## Debug long-running Recalculate Styles events in the Performance tool with new selector stats

Reviewer: I just feel like this title is too long.

```suggestion
## Debug long-running Recalculate Style events with new selector stats
```

Author: fixed locally

![Heap Snapshot Visualizer 2](./devtools-109-images/heap-snapshot-visualizer-2.png)
<!--
Instructions for screenshot
1. Get the heap snapshot visualizer extension from: Heap Snapshot visualizer - Microsoft Edge Addons
2. With the extension installed, open DevTools.
3. Go to the new Heap Snapshot Visualizer tool.
4. Choose file and select the attached .heapsnapshot file that's attached.
5. Select the Load button.
6. The Heap Snapshot Visualizer should show the heap snapshot as a graph. Refer to the attached screenshot.
7. Take a screenshot.
8. Now select the Tree button. The extension should now show the heap snapshot as a tree. Refer to the attached screenshot.
9. Take a screenshot.
-->

See also:
* [Heap Snapshot Visualizer](https://microsoftedge.microsoft.com/addons/detail/heap-snapshot-visualizer/fceldlhognbemkgfacnffkdanocidgce) - Microsoft Edge Add-ons Store.

<!-- ====================================================================== -->
## Improvements to Quick View in Focus Mode

<!-- Subtitle: Quick View now allows you to pick any tool, show several tools at once and persists across sessions. -->

> We listened to your feedback and improved the **Quick View** options in Focus Mode.  Instead of choosing from a list of predefined tools, you can now pick any tool inside DevTools and show several tools at the same time by clicking a **+** button, much like in the main toolbar of DevTools.

Reviewer 2: ```suggestion
We listened to your feedback and improved the **Quick View** options in Focus Mode.  Instead of choosing from a list of predefined tools, you can now pick any tool inside DevTools and show several tools at the same time by clicking **More tools** (`+`), much like in the main toolbar of DevTools.
```

Author: added mid-sentence icon.  recently rewrote paragraph vs the Outdated view here
